app/misc/git.ts

```diff
@@ -747,43 +749,96 @@ function pullFromRemote() {
+        // returns commit that head is pointing too (most recent local commit)
+        return repo.getHeadCommit().then(function (commit) {
+            return repo.getReferenceCommit(origin).then(function (remoteCommit) {
+                // return an object { ahead : <number of commits ahead>, behind: <number of commits behind>
+                return Graph.aheadBehind(repo, commit.id(), remoteCommit.id()).then(function (aheadBehind) {
+                    return aheadBehind;
+                });
+            }, function (e) {
+                console.log(e.message)
+            });
+        });
+    });
+}
+
+//checks if the remote version of your current branch exist
+function checkIfExistOrigin(branchName) {
+    let origin = "origin";
+    origin = path.join(origin, branchName);
+    return Git.Repository.open(repoFullPath).then(function (repo) {
+        return repo.getReferenceCommit(origin).then(function (originCommit) {
+            return true;
+        }, function (e) {
+            return false
+        });
+    });
+}
+
 function pushToRemote() {
-    if (CommitButNoPush === 0) {
-        window.alert("Cannot push without a commit.");
-        return;
-    }
-  let branch = document.getElementById("branch-name").innerText;
-  Git.Repository.open(repoFullPath)
-    .then(function (repo) {
-      console.log("Pushing changes to remote")
-      displayModal("Pushing changes to remote...");
-      addCommand("git push -u origin " + branch);
-      repo.getRemotes()
-        .then(function (remotes) {
-          repo.getRemote(remotes[0])
-            .then(function (remote) {
-              return remote.push(
-                ["refs/heads/" + branch + ":refs/heads/" + branch],
-                {
-                  callbacks: {
-                    // obtain a new copy of cred every time when user push.
-                    credentials: function () {
-                      let user = new createCredentials(getUsernameTemp(), getPasswordTemp());
-                      cred = user.credentials;
-                      return cred;
-                    }
-                  }
+    // checking status of remote repository and only push if you are ahead of remote
+    let branch = document.getElementById("branch-name").innerText;
+    //checks if the remote version of your current branch exist
+    checkIfExistOrigin(branch).then(function(remoteBranchExist){
+        if (!remoteBranchExist) {
+            window.alert("fatal: The current branch test-branch has no upstream branch.\n" +
+                "To push the current branch and set the remote as upstream, use\n" +
+                "\n" +
+                "    git push --set-upstream origin test-branch");
+            return;
+        } else {
+            // tells the user if their branch is up to date or behind the remote branch
+            getAheadBehindCommits(branch).then(function (aheadBehind) {
+                if (aheadBehind.ahead === 0) {
```

Comment:
should this if and its else statement not be switched? In the case that you are not ahead any commits locally but are still behind the remote (ie. aheadBehind.ahead = 0 and aheadBehind.behind > 0) you would want to receive the "your branch is behind remote by..." instead of "your branch is up to date"